### PR TITLE
Added support for multiple solutions in depth

### LINKED ISSUE #43: Search for all possible solutions instead of only first
**Summary.**
Search for all possible solutions instead of only first

### PR DESCRIPTION
**Pull request by DorianMoreno**
### Summary
* Now it compares the treepat expression in every depth level.
* Ascertained functionality of possible error in `ImpTargetTreeNode.matchedNodesString()`.
* Added a support grammar rule and ASTNode to represent the root of the treepat expression.
* Updated test cases that now find multiple answers.

### Test Plan
mvn test
More tests might be implemented in the future

### Related Issues
Close #43

### CODE REVIEW

File: src/main/java/functions/TreepatFunction.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package functions
+
+fun treepatFunction(expression: VisitorFunction): VisitorFunction {
+    return { targetTreeNode ->
+        if (targetTreeNode == null) {
+            VisitorFunctionResponse(listOf(VisitorFunctionSimpleResponse(lastVisitedSibling = targetTreeNode)))
+        }
```
Comment (johan-smc) [inline src/main/java/functions/TreepatFunction.kt]: It is needed? If the target tree node is null, the other code catch the invalid response

File: src/main/java/functions/TreepatFunction.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        var currentNode = targetTreeNode
+        val answers = mutableListOf<VisitorFunctionResponse>()
+        answers.add(expression.invoke(currentNode))
+        currentNode = currentNode?.nextLeftmostPreorderNode()
```
Comment (johan-smc) [inline src/main/java/functions/TreepatFunction.kt]: Delete this lines, in the while execute the same.

Comment (DorianMoreno) [reply src/main/java/functions/TreepatFunction.kt]: You're right

File: src/main/java/functions/TreepatFunction.kt (hunk 1 above)
Comment (DorianMoreno) [reply src/main/java/functions/TreepatFunction.kt]: Yes, I should get rid of that.

File: src/main/java/antlr/treepat/TreepatVisitorImplementation.kt [hunk 3]
```diff
@@ -49,7 +50,7 @@ class TreepatVisitorImplementation : TreepatVisitor<ASTNode> {
     }
 
     override fun visitIndent(ctx: TreepatParser.IndentContext): ASTNode {
-        return Check(ctx.treepat().accept(this))
+        return Check(ctx.subtree().accept(this))
```
Comment (juanpa097) [inline src/main/java/antlr/treepat/TreepatVisitorImplementation.kt]: Lets simplify the functions that just have one return by doing:
```
override fun visitIndent(ctx: TreepatParser.IndentContext): ASTNode = Check(ctx.subtree().accept(this))
```
Same for the ones bellow.

File: src/main/java/functions/TreepatFunction.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+            currentNode = currentNode.nextLeftmostPreorderNode()
+        }
+        val simpleResponses = answers.filter { it.hasMatch }.flatMap { it.responses }.filter { it.matches.isNotEmpty() }
+        if (simpleResponses.isNotEmpty()) {
+            VisitorFunctionResponse(simpleResponses, true)
+        } else {
+            VisitorFunctionResponse(listOf(VisitorFunctionSimpleResponse(lastVisitedSibling = targetTreeNode)))
+        }
+    }
+}
```
Comment (juanpa097) [inline src/main/java/functions/TreepatFunction.kt]: We can also simply by:
```suggestion
fun treepatFunction(expression: VisitorFunction): VisitorFunction = { targetTreeNode ->
        var currentNode = targetTreeNode
        val answers = mutableListOf<VisitorFunctionResponse>()
        while (currentNode != null) {
            answers.add(expression.invoke(currentNode))
            currentNode = currentNode.nextLeftmostPreorderNode()
        }
        val simpleResponses = answers.filter { it.hasMatch }.flatMap { it.responses }.filter { it.matches.isNotEmpty() }
        if (simpleResponses.isNotEmpty()) {
            VisitorFunctionResponse(simpleResponses, true)
        } else {
            VisitorFunctionResponse(listOf(VisitorFunctionSimpleResponse(lastVisitedSibling = targetTreeNode)))
        }
    }

```

File: src/main/java/antlr/treepat/TreepatVisitorImplementation.kt (hunk 3 above)
Comment (DorianMoreno) [reply src/main/java/antlr/treepat/TreepatVisitorImplementation.kt]: Ok, I also orginized the file so all one line functions are together.

File: src/main/java/functions/TreepatFunction.kt (hunk 4 above)
Comment (DorianMoreno) [reply src/main/java/functions/TreepatFunction.kt]: Fine, will get to it

**Pull request merged**